message/infrastructure/http/publisher.go

```diff
@@ -0,0 +1,104 @@
+package http
+
+import (
+	"bytes"
+	"net/http"
+
+	"github.com/pkg/errors"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+)
+
+var (
+	// ErrPublisherClosed happens when trying to publish to a topic while the publisher is closed or closing.
+	ErrPublisherClosed = errors.New("publisher is closed")
+)
+
+type MarshalMessageFunc func(topic string, msg *message.Message) (*http.Request, error)
+
+// DefaultMarshalMessageFunc encodes the UUID and Metadata in request headers
+// and sets the request URL according to the predefined server address and the topic.
+func DefaultMarshalMessageFunc(address string) MarshalMessageFunc {
+	return func(topic string, msg *message.Message) (*http.Request, error) {
+		req, err := http.NewRequest(http.MethodPost, address+"/"+topic, bytes.NewBuffer(msg.Payload))
+		if err != nil {
+			return nil, err
+		}
+
+		req.Header.Set(HeaderUUID, msg.UUID)
+
+		metadataJson, err := metadataToJson(msg)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not marshal metadata to JSON")
+		}
+		req.Header.Set(HeaderMetadata, string(metadataJson))
+		return req, nil
+	}
+}
+
+type Publisher struct {
+	client *http.Client
+	logger watermill.LoggerAdapter
+
+	marshalMessageFunc MarshalMessageFunc
+
+	closed bool
+}
+
+func NewPublisher(marshalMessageFunc MarshalMessageFunc, logger watermill.LoggerAdapter) (*Publisher, error) {
+	return NewPublisherWithClient(http.DefaultClient, marshalMessageFunc, logger)
+}
+
+func NewPublisherWithClient(client *http.Client, marshalMessageFunc MarshalMessageFunc, logger watermill.LoggerAdapter) (*Publisher, error) {
+	return &Publisher{
+		client:             client,
+		logger:             logger,
+		marshalMessageFunc: marshalMessageFunc,
+	}, nil
+}
+
+func (p *Publisher) Publish(topic string, messages ...*message.Message) error {
+	if p.closed {
+		return ErrPublisherClosed
+	}
+
+	for _, msg := range messages {
+		req, err := p.marshalMessageFunc(topic, msg)
+		if err != nil {
+			return errors.Wrapf(err, "cannot marshal message %s", msg.UUID)
+		}
+
+		logFields := watermill.LogFields{
+			"uuid":     msg.UUID,
+			"url":      req.URL.String(),
+			"method":   req.Method,
+			"provider": ProviderName,
+		}
+
+		resp, err := p.client.Do(req)
+		if err != nil {
+			return errors.Wrapf(err, "publishing message %s failed", msg.UUID)
+		}
+
+		// todo: process the response anyhow?
```

Comment:
a9666ef7fec40fcf8c6a774ae04a87e059e46caa ok, now I'm quite satisfied with how errors are handled from this request. Also, I made sure that `resp.Body.Close()` is always called to prevent leaks, even when there was no server error.